clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -331,31 +332,59 @@ else if (error.exception() instanceof RetriableException)
      */
     private void resetOffset(TopicPartition partition) {
         OffsetResetStrategy strategy = subscriptions.resetStrategy(partition);
-        final long timestamp;
-        if (strategy == OffsetResetStrategy.EARLIEST)
-            timestamp = ListOffsetRequest.EARLIEST_TIMESTAMP;
-        else if (strategy == OffsetResetStrategy.LATEST)
-            timestamp = ListOffsetRequest.LATEST_TIMESTAMP;
-        else
-            throw new NoOffsetForPartitionException(partition);
-
         log.debug("Resetting offset for partition {} to {} offset.", partition, strategy.name().toLowerCase(Locale.ROOT));
-        long offset = getOffsetsByTimes(Collections.singletonMap(partition, timestamp), Long.MAX_VALUE).get(partition).offset();
-
+        long offset = listOffset(partition, strategy);
         // we might lose the assignment while fetching the offset, so check it is still active
         if (subscriptions.isAssigned(partition))
             this.subscriptions.seek(partition, offset);
     }
 
+    private long listOffset(TopicPartition partition, OffsetResetStrategy strategy) {
+        final long timestamp;
+        switch (strategy) {
+            case EARLIEST:
+                timestamp = ListOffsetRequest.EARLIEST_TIMESTAMP;
+                break;
+            case LATEST:
+                timestamp = ListOffsetRequest.LATEST_TIMESTAMP;
+                break;
+            default:
+                throw new NoOffsetForPartitionException(partition);
+        }
+        while (true) {
+            RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> future =
+                    sendListOffsetRequests(false, Collections.singletonMap(partition, timestamp));
+            client.poll(future);
+            if (future.succeeded()) {
+                Map<TopicPartition, OffsetAndTimestamp> ret = future.value();
+                return ret.get(partition).offset();
```

Comment:
Hmm, I didn't see any code path where the return value could be null, or the value of the future itself, or the keys or values of the map.  It seems like if any of those things were to happen it would be a programming error where an NPE is appropriate?  Unless I missed something here